frame/treasury/src/lib.rs

```diff
@@ -625,6 +786,333 @@ decl_module! {
+					curator: signer,
+					beneficiary: beneficiary.clone(),
+					unlock_at: system::Module::<T>::block_number() + T::BountyDepositPayoutDelay::get(),
+				};
+
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::BountyAwarded(bounty_id, beneficiary));
+		}
+
+		/// Claim the payout from an awarded bounty after payout delay.
+		///
+		/// The dispatch origin for this call must be the beneficiary of this bounty.
+		///
+		/// - `bounty_id`: Bounty ID to claim.
+		#[weight = T::WeightInfo::claim_bounty()]
+		fn claim_bounty(origin, #[compact] bounty_id: BountyIndex) {
+			let _ = ensure_signed(origin)?; // anyone can trigger claim
+
+			Bounties::<T>::try_mutate_exists(bounty_id, |maybe_bounty| -> DispatchResult {
+				let bounty = maybe_bounty.take().ok_or(Error::<T>::InvalidIndex)?;
+				if let BountyStatus::PendingPayout { curator, beneficiary, unlock_at } = bounty.status {
+					ensure!(system::Module::<T>::block_number() >= unlock_at, Error::<T>::Premature);
+					let bounty_account = Self::bounty_account_id(bounty_id);
+					let balance = T::Currency::free_balance(&bounty_account);
+					let fee = bounty.fee;
+					let payout = balance.saturating_sub(fee);
+					let _ = T::Currency::unreserve(&curator, bounty.curator_deposit);
+					let _ = T::Currency::transfer(&bounty_account, &curator, fee, AllowDeath); // should not fail
+					let _ = T::Currency::transfer(&bounty_account, &beneficiary, payout, AllowDeath); // should not fail
+					*maybe_bounty = None;
+
+					BountyDescriptions::remove(bounty_id);
+
+					Self::deposit_event(Event::<T>::BountyClaimed(bounty_id, payout, beneficiary));
+					Ok(())
+				} else {
+					Err(Error::<T>::UnexpectedStatus.into())
+				}
+			})?;
+		}
+
+		/// Cancel an active bounty. All the funds will be send to treasury.
+		///
+		/// Only curator is able to cancel active bounty.
+		/// Otherwise `T::RejectOrigin` is able to cancel funded bounty.
+		///
+		/// - `bounty_id`: Bounty ID to cancel.
+		#[weight = T::WeightInfo::cancel_bounty()]
+		fn cancel_bounty(origin, #[compact] bounty_id: BountyIndex) {
+			Bounties::<T>::try_mutate_exists(bounty_id, |maybe_bounty| -> DispatchResult {
+				let bounty = maybe_bounty.as_ref().ok_or(Error::<T>::InvalidIndex)?;
+
+				match &bounty.status {
+					BountyStatus::Funded |
+					BountyStatus::CuratorAssigned { .. } => {
+						T::RejectOrigin::ensure_origin(origin)?;
+					},
+					BountyStatus::PendingPayout { curator, .. } => {
+						T::RejectOrigin::ensure_origin(origin)?;
+
+						let _ = T::Currency::unreserve(curator, bounty.curator_deposit);
+					},
+					BountyStatus::Active { curator, expires } => {
+						let signer = ensure_signed(origin)?;
+
+						let now = system::Module::<T>::block_number();
+						if *expires > now {
+							// only curator can cancel unexpired bounty
+							ensure!(signer == *curator, Error::<T>::RequireCurator);
+						}
+
+						let _ = T::Currency::unreserve(curator, bounty.curator_deposit);
```

Comment:
When a curator cancels a bounty that is active, they get `unreserve`. If the `unassign_curator`, they are slashed.

They should both be `unreserve`, and the slash in `unassign_curator` should only come from the rejection origin

Reply:
On second thought, probably should be slashed here too. If a curator accepts the task, they should be on the hook for completing it. If we want to refund someone, we can do a tip.

Reply:
The point is, unassign curator means the bounty is still valid, but curator failed to perform the work, so we slash curator and looking for another one.

Cancel bounty means the bounty itself is no longer valid for whatever reason, it is not curator's fault so no slash just unreserve.

Curator can unassign himself but that means he failed to keep the promise so slash. When accepting curator role, curator is making a commitment that he will do the job.

Reply:
It seems to me curator should not have the power to cancel a bounty then, only council. Why should a curator say that a bounty is no longer needed?

Why should they be able to get their full reserve back after failing to complete the task after the expiry period has ended?